### DISCUSSION TITLE
Background thread feeding data

### DISCUSSION THREAD

**Discussion by BigfootN** [Relm4/Relm4#633] [Q&A] [answered]
Hello,

I wish to know the best way to implement this.

I have a thread with an infinite loop. This loop keeps fetching data continuously. The thread receives all kinds of data, and needs to dispatch it to the sub components/widgets.

I see two options (maybe there's a better way):
- Create the thread separately from the application and send (from the top/root component/widget) to the concerned sub component/widget
- Divide the thread in several threads. Each thread would be placed in each component,  feeding the component with the fetched data and not worrying about how to transfer data from "up top" to the "bottom"

Hoping I was clear enough.

Thank you very much in advance for any help

**Comment by AaronErhardt** [comment]
Relm4 has several utilities for working with threads or async code that should help you: https://relm4.org/book/stable/threads_and_async/index.html

In your case, I think synchronous commands are the best solution.

**Reply by BigfootN** [reply]
Thank you for your quick response and help.

So if I understand well, the idea would be to have the `update_cmd` implemented at the "root" window and pass it along to the concerned sub component (N levels below) ?

Thank you for your time and help

**Reply by AaronErhardt** [reply, marked as answer]
> So if I understand well, the idea would be to have the update_cmd implemented at the "root" window and pass it along to the concerned sub component (N levels below) ?

No, you can implement the `update_cmd` method for the component for which you need it. The basic idea is to:

1. Spawn your loop thread as command using [`spawn_command`](https://docs.rs/relm4/latest/relm4/struct.ComponentSender.html#method.spawn_command)
2. Every time you receive data, send it as a command output message to your component
3. Process the command output message in `update_cmd`

**Reply by BigfootN** [reply]
Thank you very much. Exactly what I was  looking for
